src/remote-editor-service.ts

```diff
@@ -0,0 +1,268 @@
+ *
+ * This class runs in-process and communicates via SelfChannel with
+ * EditorServer, which runs in the child process.
+ */
+export class RemoteEditorService extends BaseEditor {
+  private _channel = new SelfChannel();
+  constructor(basedir: string) {
+    super();
+    this._channel.request({kind: 'init', basedir});
+  }
+  async getWarningsFor(localPath: string): Promise<Warning[]> {
+    return this._channel.request({kind: 'getWarningsFor', localPath});
+  }
+  async fileChanged(localPath: string, contents?: string): Promise<void> {
+    return this._channel.request({kind: 'fileChanged', localPath, contents});
+  }
+
+  async getDocumentationFor(localPath: string, position: Position):
+      Promise<string|undefined> {
+    return this._channel.request(
+        {kind: 'getDocumentationFor', localPath, position});
+  }
+
+  async getDefinitionFor(localPath: string, position: Position):
+      Promise<SourceRange> {
+    return this._channel.request(
+        {kind: 'getDefinitionFor', localPath, position});
+  }
+
+  async getTypeaheadCompletionsFor(localPath: string, position: Position):
+      Promise<TypeaheadCompletion|undefined> {
+    return this._channel.request(
+        {kind: 'getTypeaheadCompletionsFor', localPath, position});
+  }
+
+  async clearCaches(): Promise<void> {
+    return this._channel.request({kind: 'clearCaches'});
+  }
+
+  dispose(): void {
+    this._channel.dispose();
+  }
+}
+
+/**
+ * Runs out of process and handles
+ */
+class EditorServer {
+  private _editorService: EditorService;
+  constructor(basedir: string) {
+    this._editorService = new EditorService({
+      urlLoader: new FSUrlLoader(basedir),
+      urlResolver: new PackageUrlResolver()
+    });
+  }
+
+  async handleMessage(message: Request): Promise<any> {
+    switch (message.kind) {
+      case 'getWarningsFor':
+        return this._editorService.getWarningsFor(message.localPath);
+      case 'fileChanged':
+        await this._editorService.fileChanged(
+            message.localPath, message.contents);
+        return;
+      case 'init':
+        throw new Error('Already initialized!');
+      case 'getDefinitionFor':
+        return this._editorService.getDefinitionFor(
+            message.localPath, message.position);
+      case 'getDocumentationFor':
+        return this._editorService.getDocumentationFor(
+            message.localPath, message.position);
+      case 'getTypeaheadCompletionsFor':
+        return this._editorService.getTypeaheadCompletionsFor(
+            message.localPath, message.position);
+      case 'clearCaches':
+        return this._editorService.clearCaches();
+      default:
+        // This assignment makes it a type error if we don't handle all possible
+        // values of `message.kind`.
+        const never: never = message;
+        throw new Error(`Got unknown kind of message: ${util.inspect(never)}`);
+    }
+  }
+}
+
+function makeDeferred<V>(): Deferred<V> {
```

Comment:
The class will probably be smaller than the interface + function, and it'll be in one place, making it easier to read. I bet we'll find uses for it elsewhere too.